MagazineLayout/LayoutCore/SectionModel.swift

```diff
@@ -223,112 +208,113 @@ struct SectionModel {
-    headerModel = nil
+    if let indexOfHeader = indexOfHeaderRow() {
+      updateIndexOfFirstInvalidatedRowIfNecessary(toProposedIndex: indexOfHeader)
+    }
 
-    indexOfFirstInvalidatedItem = 0
+    headerModel = nil
   }
 
   mutating func removeFooter() {
-    footerModel = nil
+    if let indexOfFooter = indexOfFooterRow() {
+      updateIndexOfFirstInvalidatedRowIfNecessary(toProposedIndex: indexOfFooter)
+    }
 
-    indexOfFirstInvalidatedItem = 0
+    footerModel = nil
   }
 
   mutating func updateHeaderHeight(toPreferredHeight preferredHeight: CGFloat) {
     headerModel?.preferredHeight = preferredHeight
 
-    indexOfFirstInvalidatedItem = 0
+    if let indexOfHeader = indexOfHeaderRow() {
+      updateIndexOfFirstInvalidatedRowIfNecessary(toProposedIndex: indexOfHeader)
+    }
   }
 
   mutating func updateFooterHeight(toPreferredHeight preferredHeight: CGFloat) {
     footerModel?.preferredHeight = preferredHeight
 
-    indexOfFirstInvalidatedItem = 0
+    if let indexOfFooter = indexOfFooterRow() {
+      updateIndexOfFirstInvalidatedRowIfNecessary(toProposedIndex: indexOfFooter)
+    }
   }
 
   mutating func setBackground(_ backgroundModel: BackgroundModel) {
     self.backgroundModel = backgroundModel
 
-    indexOfFirstInvalidatedItem = 0
+    let indexOfLastRow = indexOfFooterRow() ?? indexOfLastRowOfItems() ?? indexOfHeaderRow() ?? -1
+    updateIndexOfFirstInvalidatedRowIfNecessary(toProposedIndex: indexOfLastRow + 1)
   }
 
   mutating func removeBackground() {
     backgroundModel = nil
+    // No need to invalidate since no frames will be adjusted
   }
 
   // MARK: Private
 
   private var itemModels: [ItemModel]
   private var metrics: MagazineLayoutSectionMetrics
   private var calculatedHeight: CGFloat
-  private var indexOfFirstInvalidatedItem: Int?
-
-  private mutating func updateIndexOfFirstInvalidatedItem(
-    forChangeToItemAtIndex changedIndex: Int)
-  {
-    guard changedIndex > 0 else {
-      indexOfFirstInvalidatedItem = changedIndex
-      return
-    }
 
-    let candidateFirstAffectedIndex = min(itemModels.count - 1, changedIndex)
+  private var indexOfFirstInvalidatedRow: Int?
+  private var itemIndicesForRowIndices = [Int: [Int]]()
+  private var rowIndicesForItemIndices = [Int: Int]()
 
-    let indexRangeOfItemsInSameRow = indexRangeOfItemsInSameRowAsItem(
-      atIndex: candidateFirstAffectedIndex)
-    var rowContainsStretchToTallestItemInRowItem = false
-    for index in indexRangeOfItemsInSameRow {
-      if itemModels[index].sizeMode.heightMode == .dynamicAndStretchToTallestItemInRow {
-        rowContainsStretchToTallestItemInRowItem = true
-        break
-      }
+  private mutating func updateIndexOfFirstInvalidatedRow(forChangeToItemAtIndex changedIndex: Int) {
+    guard
+      let indexOfCurrentRow = rowIndicesForItemIndices[changedIndex],
+      indexOfCurrentRow > 0 else
+    {
+      indexOfFirstInvalidatedRow = rowIndicesForItemIndices[0] ?? 0
```

Comment:
`rowIndicesForItemIndices.first`?

Reply:
This is a dictionary of row indices keyed by item indices, so I actually do need to grab the value for item index 0 as the key.